File: array/136-Single-Number.js
/*
136. Single Number
https://leetcode.com/problems/single-number/
18 June 2023
David Pratt
*/

/**
 * @param {number[]} nums
 * @return {number}
 */

var singleNumber = function(nums) {
    let map = {}

    // count the number of occurrences
    for(let i = 0; i < nums.length; i++){
        if(map[nums[i]]) map[ nums[i] ] += 1;
        else map[nums[i]] = 1;
    }

    let entries = Object.entries(map);
    let filteredEntries = entries.filter(x => x[1] === 1)
    return Number(filteredEntries[0][0])

};

File: array/136-Single-Number.py
/*
136. Single Number
https://leetcode.com/problems/single-number/
26 November 2023
David Pratt


*/

class Solution:
    def singleNumber(self, nums: List[int]) -> int:
        numCount = {};
        for num in nums:
            if num in numCount:
                numCount[num] += 1
            else:
                numCount[num] = 1
        
        for key, value in numCount.items():
            if value == 1:
                return key
        
        return -1
